stan/math/rev/mat/fun/cholesky_decompose.hpp

```diff
@@ -223,7 +235,157 @@ class cholesky_scalar : public vari {
+    block_size_ = std::max(block_size_, 8);
+    block_size_ = std::min(
+        block_size_, opencl_context.tuning_opts().cholesky_rev_min_block_size);
+    // The following is a GPU implementation of
+    // the chain() function from the cholesky_block
+    // vari class implementation
+    for (int k = M_; k > 0; k -= block_size_) {
+      const int j = std::max(0, k - block_size_);
+      const int k_j_ind = k - j;
+      const int m_k_ind = M_ - k;
+
+      matrix_gpu R(k_j_ind, j);
+      matrix_gpu D(k_j_ind, k_j_ind);
+      matrix_gpu B(m_k_ind, j);
+      matrix_gpu C(m_k_ind, k_j_ind);
+
+      matrix_gpu Rbar(k_j_ind, j);
+      matrix_gpu Dbar(k_j_ind, k_j_ind);
+      matrix_gpu Bbar(m_k_ind, j);
+      matrix_gpu Cbar(m_k_ind, k_j_ind);
+
+      R.sub_block(L, j, 0, 0, 0, k_j_ind, j);
+      D.sub_block(L, j, j, 0, 0, k_j_ind, k_j_ind);
+      B.sub_block(L, k, 0, 0, 0, m_k_ind, j);
+      C.sub_block(L, k, j, 0, 0, m_k_ind, k_j_ind);
+
+      Rbar.sub_block(Lbar, j, 0, 0, 0, k_j_ind, j);
+      Dbar.sub_block(Lbar, j, j, 0, 0, k_j_ind, k_j_ind);
+      Bbar.sub_block(Lbar, k, 0, 0, 0, m_k_ind, j);
+      Cbar.sub_block(Lbar, k, j, 0, 0, m_k_ind, k_j_ind);
+
+      Cbar = Cbar * lower_triangular_inverse(D);
+      Bbar = Bbar - Cbar * R;
+      Dbar = Dbar - transpose(Cbar) * C;
+
+      // the implementation of the symbolic_rev inline function
+      // for the GPU
+      Dbar = transpose(D) * Dbar;
+      Dbar.triangular_transpose<TriangularMapGPU::LowerToUpper>();
+      D = transpose(lower_triangular_inverse(D));
+      Dbar = D * transpose(D * Dbar);
+      Dbar.triangular_transpose<TriangularMapGPU::LowerToUpper>();
+      // end of symbolic_rev inline function
+
+      Rbar = Rbar - transpose(Cbar) * B;
+      Rbar = Rbar - Dbar * R;
+      Dbar = diagonal_multiply(Dbar, 0.5);
+      Dbar.zeros<stan::math::TriangularViewGPU::Upper>();
+
+      Lbar.sub_block(Rbar, 0, 0, j, 0, k_j_ind, j);
+      Lbar.sub_block(Dbar, 0, 0, j, j, k_j_ind, k_j_ind);
+      Lbar.sub_block(Bbar, 0, 0, k, 0, m_k_ind, j);
+      Lbar.sub_block(Cbar, 0, 0, k, j, m_k_ind, k_j_ind);
+    }
+    copy(Lbar_, Lbar);
+    pos = 0;
+    for (size_type j = 0; j < M_; ++j)
+      for (size_type i = j; i < M_; ++i)
+        variRefA_[pos++]->adj_ += Lbar_.coeffRef(i, j);
+  }
+};
+#endif
 
+/**
+ * Set the lower right triangular of a var matrix given a set of vari**
+ *
+ * @param L Matrix of vars
+ * @param variRef Values to be set in lower right triangular of L.
+ * @return Matrix of vars
+ */
+inline Eigen::Matrix<var, -1, -1>
+ set_lower_tri_coeff_ref(Eigen::Matrix<var, -1, -1>& L, vari**& variRef) {
+  size_t pos = 0;
+  vari* dummy = new vari(0.0, false);
+
+  for (size_type j = 0; j < L.cols(); ++j) {
+    for (size_type i = j; i < L.cols(); ++i) {
+      L.coeffRef(i, j).vi_ = variRef[pos++];
+    }
+    for (size_type k = 0; k < j; ++k)
+      L.coeffRef(k, j).vi_ = dummy;
+  }
+  return L;
+}
```

Comment:
@seantalts I like this because it cleans up the code below a lot, but I'm not really sure where to put it.

 Good, bad, thoughts?

Also not sure if it should return `Matrix<var, -1, -1>` or `void` since it's manipulating a reference to L

Reply:
In most of the communities I've been in, it's considered good style to either have a function be pure and return the result, or return `void` and have the inputs clearly marked for modification. So here I would return void. 

Might be good to put it in `internal` - I think it's fine in this file.

Reply:
:+1 agree on void I'll put this in `internal`